### PR TITLE
[Fleet] Add source information to configuration settings

### PR DESCRIPTION
**Pull request by julesmcrt**
Previous PR before packages got moved around : https://github.com/DataDog/datadog-agent/pull/19832

### TL;DR
Know which source is setting a configuration parameter

### Future consequences
When using the `.Set()` method to change a configuration setting in the agent, you must now add a `model.Source` parameter, keeping in mind the following hierarchy between them:
`CLI > Remote Config > Agent Runtime > Env Var > File > Unknown > Default`

You could also use the `.SetWithoutSource()` method, it will then assume `SourceUnknown`

If you want to override a setting (e.g. for a unit test), you can now call `.Set()` specifying a source higher in the hierarchy, then use the `.UnsetForSource()` method ; the applied setting will fallback to its previous value

### What do the sources correspond to?

1. `Default` is the lowest in the hierarchy, also Go fallback state (empty string). Also used when calling `.SetDefault()`
2. `Unknown` is used by the `.SetWithoutSource()` method, you should avoid using it on purpose
3. `File` corresponds to the `datadog.yaml` configuration file
4. `Env Var` corresponds to the environment variables
5. `Agent Runtime` is used when the agent itself changes values for its configuration, **not to be confused with runtime settings**
6. `Remote Config` is used to remotely change the behaviour of the agent
7. `CLI` is used when a user manually enters commands

### What does this PR do?

1. Move the `Source` type from the `settings` package to `model` (Viper wrapper)
8. Add one Viper instance per source, and edit the `.Set()` method to include a `source` parameter
9. Add a `.SetWithoutSource()` as a legacy way to change a setting without specifying a source
10. Add the different sources to the inventories payload

### Motivation

For Fleet Automation, we want to display the configuration of the agent and the source of every setting.
Use case: a customer doesn't understand why when he changed a setting in the `datadog.yaml` file it didn't have any effect on the agent (because he had forgotten to unset an `ENV VAR`).

### Additional Notes
A lot of the changes are moving `.Set()` calls to `.SetWithoutSource()`, the more impactful changes are in the following files:

1. pkg/config/model/viper.go
2. Adding the sources to the Agent inventories payload
3. comp/remote-config/rcclient/rcclient.go

### More detailed explanations 
Because Viper doesn't support multiple sources of setting, the workaround is to spawn one instance of Viper per configuration source. Everytime a setting is changed, we check the value of every instance and the value of the highest source is applied to the "main" Viper instance.

Only the `Set` core concept is changed, the `Get` behavior do not change

### Possible Drawbacks / Trade-offs

1. The new Viper instances will require more computing for the Agent
5. The inventories payload will be heavier, REDAPL agreed for the storage aspect

### Describe how to test/QA your changes

- Ensure that functions which change agent configuration settings still behave as expected
- Use different sources to change settings and ensure that the hierarchy is always respected
- Check CPU & memory usage, especially if you use Runtime Settings or Remote Config

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: comp/remote-config/rcclient/rcclient.go [hunk 1]
```diff
@@ -121,59 +114,38 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
...
-			pkglog.Infof("Removing remote-config log level override, falling back to '%s'", newLevel)
+			pkglog.Infof("Removing remote-config log level override, falling back to '%s'", config.Datadog.Get("log_level"))
+			config.Datadog.UnsetForSource("log_level", model.SourceRC)
 		} else {
-			newLevel = mergedConfig.LogLevel
-			newSource = settings.SourceRC
+			newLevel := mergedConfig.LogLevel
+			newSource := model.SourceRC
 			pkglog.Infof("Changing log level to '%s' through remote config", newLevel)
+			err = settings.SetRuntimeSetting("log_level", newLevel, newSource)
```
Comment (BaptisteFoy) [inline comp/remote-config/rcclient/rcclient.go:131]: **🥜 Nitpick:** ‏We can definitely simplify some variables here

File: comp/trace/config/config.go [hunk 2]
```diff
@@ -89,7 +89,7 @@ func (c *cfg) SetHandler() http.Handler {
 					httpError(w, http.StatusInternalServerError, err)
 					return
 				}
-				pkgconfig.Datadog.Set("log_level", lvl)
+				pkgconfig.Datadog.SetWithoutSource("log_level", lvl)
```
Comment (BaptisteFoy) [inline comp/trace/config/config.go:92]: Let's set a source

File: comp/trace/config/setup.go [hunk 3]
```diff
@@ -845,7 +845,7 @@ func SetHandler() http.Handler {
 					httpError(w, http.StatusInternalServerError, err)
 					return
 				}
-				coreconfig.Datadog.Set("log_level", lvl)
+				coreconfig.Datadog.SetWithoutSource("log_level", lvl)
```
Comment (BaptisteFoy) [inline comp/trace/config/setup.go:848]: Let's set a source

File: pkg/aggregator/demultiplexer_agent.go [hunk 4]
```diff
@@ -152,7 +152,7 @@ func initAgentDemultiplexer(log log.Component, sharedForwarder forwarder.Forward
 
 	if config.Datadog.GetBool("telemetry.enabled") && config.Datadog.GetBool("telemetry.dogstatsd_origin") && !config.Datadog.GetBool("aggregator_use_tags_store") {
 		log.Warn("DogStatsD origin telemetry is not supported when aggregator_use_tags_store is disabled.")
-		config.Datadog.Set("telemetry.dogstatsd_origin", false)
+		config.Datadog.SetWithoutSource("telemetry.dogstatsd_origin", false)
```
Comment (BaptisteFoy) [inline pkg/aggregator/demultiplexer_agent.go:155]: Let's set a source

File: pkg/cli/subcommands/check/command.go [hunk 5]
```diff
@@ -183,8 +183,8 @@ func run(config config.Component, cliParams *cliParams, demultiplexer demultiple
 		if pkgconfig.Datadog.IsSet("integration_tracing_exhaustive") {
 			previousIntegrationTracingExhaustive = pkgconfig.Datadog.GetBool("integration_tracing_exhaustive")
 		}
-		pkgconfig.Datadog.Set("integration_tracing", true)
-		pkgconfig.Datadog.Set("integration_tracing_exhaustive", true)
+		pkgconfig.Datadog.SetWithoutSource("integration_tracing", true)
+		pkgconfig.Datadog.SetWithoutSource("integration_tracing_exhaustive", true)
```
Comment (BaptisteFoy) [inline pkg/cli/subcommands/check/command.go:187]: Let's set a source

File: pkg/cli/subcommands/check/command.go [hunk 6]
```diff
@@ -497,8 +497,8 @@ func run(config config.Component, cliParams *cliParams, demultiplexer demultiple
 	}
 
 	if cliParams.generateIntegrationTraces {
-		pkgconfig.Datadog.Set("integration_tracing", previousIntegrationTracing)
-		pkgconfig.Datadog.Set("integration_tracing_exhaustive", previousIntegrationTracingExhaustive)
+		pkgconfig.Datadog.SetWithoutSource("integration_tracing", previousIntegrationTracing)
+		pkgconfig.Datadog.SetWithoutSource("integration_tracing_exhaustive", previousIntegrationTracingExhaustive)
```
Comment (BaptisteFoy) [inline pkg/cli/subcommands/check/command.go:501]: Let's set a source

File: pkg/config/process.go [hunk 7]
```diff
@@ -217,14 +217,14 @@ func loadProcessTransforms(config Config) {
...
 		} else if enabled, _ := strconv.ParseBool(procConfigEnabled); enabled { // "true"
-			config.Set("process_config.process_collection.enabled", true)
-			config.Set("process_config.container_collection.enabled", false)
+			config.SetWithoutSource("process_config.process_collection.enabled", true)
+			config.SetWithoutSource("process_config.container_collection.enabled", false)
 		} else { // "false"
-			config.Set("process_config.process_collection.enabled", false)
-			config.Set("process_config.container_collection.enabled", true)
+			config.SetWithoutSource("process_config.process_collection.enabled", false)
+			config.SetWithoutSource("process_config.container_collection.enabled", true)
```
Comment (BaptisteFoy) [inline pkg/config/process.go:227]: Let's set a source

File: pkg/serverless/metrics/metric.go [hunk 8]
```diff
@@ -65,17 +65,17 @@ func (m *MetricDogStatsD) NewServer(demux aggregator.Demultiplexer) (dogstatsdSe
...
 	// we want to avoid duplicate metric
 	customerList := config.Datadog.GetStringSlice(statsDMetricBlocklistKey)
 
 	// if the proxy is enabled we need to also block the errorMetric
 	if strings.ToLower(os.Getenv(proxyEnabledEnvVar)) == "true" {
-		config.Datadog.Set(statsDMetricBlocklistKey, buildMetricBlocklistForProxy(customerList))
+		config.Datadog.SetWithoutSource(statsDMetricBlocklistKey, buildMetricBlocklistForProxy(customerList))
 	} else {
-		config.Datadog.Set(statsDMetricBlocklistKey, buildMetricBlocklist(customerList))
+		config.Datadog.SetWithoutSource(statsDMetricBlocklistKey, buildMetricBlocklist(customerList))
```
Comment (BaptisteFoy) [inline pkg/serverless/metrics/metric.go:78]: Let's set a source

File: pkg/serverless/trace/trace.go [hunk 9]
```diff
@@ -83,7 +83,7 @@ func (s *ServerlessTraceAgent) Start(enabled bool, loadConfig Load, lambdaSpanCh
 		// Set the serverless config option which will be used to determine if
 		// hostname should be resolved. Skipping hostname resolution saves >1s
 		// in load time between gRPC calls and agent commands.
-		ddConfig.Datadog.Set("serverless.enabled", true)
+		ddConfig.Datadog.SetWithoutSource("serverless.enabled", true)
```
Comment (BaptisteFoy) [inline pkg/serverless/trace/trace.go:86]: Let's set a source

**Comment by julesmcrt** [comment]
Following @BaptisteFoy review: the `.SetWithoutSource()` method is only used in tests and mocks

**Review by BaptisteFoy — APPROVED** [review]
LGTM for RC

**Review by hmahmood — APPROVED** [review]

**Review by NouhaManai96 — APPROVED** [review]
looks good for NDM

**Review by chouquette — APPROVED** [review]
LGTM for agent-platform owned files

**Review by paulcacheux — APPROVED** [review]

**Review by purple4reina — APPROVED** [review]

**Review by YoannGh — APPROVED** [review]

**Review by songy23 — APPROVED** [review]
LGTM on changes to OTel

**Review by brycekahle — APPROVED** [review]

**Review by liustanley — APPROVED** [review]

**Review by derekwbrown — APPROVED** [review]

**Review by sarah-witt — APPROVED** [review]
Looks good for platform integrations!

**Review by guyarb — APPROVED** [review]

**Review by dustmop — APPROVED** [review]
Looks good for Agent Shared Components

**Review by clamoriniere — APPROVED** [review]

**Review by paullegranddc — APPROVED** [review]

**Review by knusbaum — APPROVED** [review]

**Review by mfpierre — APPROVED** [review]

**Review by vickenty — APPROVED** [review]

**Review by xornivore — APPROVED** [review]
